File: workout-api/src/main/java/tom/projects/monkeylog/repository/workout/WorkoutRepository.java
package tom.projects.monkeylog.repository.workout;

import org.springframework.data.jpa.repository.JpaRepository;
import tom.projects.monkeylog.model.workout.Workout;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

public interface WorkoutRepository extends JpaRepository<Workout, Long> {
    List<Workout> findAllByTypeAndUserId(Workout.Type type, Long userId);

    Optional<Workout> findByIdAndType(Long id, Workout.Type type);

    List<Workout> findByEndDateBetween(LocalDateTime start, LocalDateTime end);
}

File: backend/bin/main/tom/com/monkeylog/repository/workout/WorkoutRepository.kt
package tom.com.monkeylog.repository.workout

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.EntityGraph
import org.springframework.data.jpa.repository.JpaRepository
import tom.com.monkeylog.model.workout.Workout
import tom.com.monkeylog.model.workout.WorkoutType
import java.time.Instant
import java.util.*

interface WorkoutRepository : JpaRepository<Workout, UUID> {
    @EntityGraph(value = "workout.exerciseGroups.exerciseRows.exerciseRowFields")
    override fun findById(id: UUID): Optional<Workout>

    @EntityGraph(value = "workout.exerciseGroups.exerciseRows.exerciseRowFields")
    fun findAllByWorkoutTypeAndUserId(workoutType: WorkoutType, userId: UUID): List<Workout>

    @EntityGraph(value = "workout.exerciseGroups.exerciseRows.exerciseRowFields")
    fun findAllByWorkoutTypeAndUserId(workoutType: WorkoutType, userId: UUID, pageable: Pageable): Page<Workout>

    @EntityGraph(value = "workout.exerciseGroups.exerciseRows.exerciseRowFields")
    fun findAllByWorkoutTypeAndUserIdAndStartDateAfter(
        workoutType: WorkoutType,
        userId: UUID,
        start: Instant,
        pageable: Pageable
    ): Page<Workout>
}
